### ISSUE TITLE
DRY data munging in Genomics : application to Variant Calling

### ISSUE THREAD

**Issue by radaniba** [minisciencegirl/studyGroup#18]
_this is a course draft, no dates are fixed yet but this will change soon, comments are welcome_
## Do-Not-Repeat-Yourself

In this 3 parts hacky hours we will learn how to get from raw data to information. We will develop a standard pipeline in Python that we can run locally as well as on a cluster for more robustness.

Our hacky hour is applied to a specific domain, but the principles are transferrable to any other field.
## What will we be doing ?

We will be managing raw data, DNA out of [sequencers](http://en.wikipedia.org/wiki/DNA_sequencer) in Fastq format. For simplicity sake we will be using [Illumina MiSeq](http://www.illumina.com/content/dam/illumina-marketing/documents/products/illumina_sequencing_introduction.pdf) data, because usually they are smaller and are more suitable for tutorial/hack day

We will learn how to :

_(week) DAY 1_
- write a python pipeline using Ruffus
- wrap publically available tools
- code versioning using git

_(week) DAY 2_
- developing a pipeline to find somatic mutations in cancer DNA
- scaling the pipeline for cluster use (support high volume of data)

_(week) DAY 3_
- Plotting the data using matplotlib
- Reproduce plots using R's ggplot2

Bonus
- Using Docker to ship our code/results
- Public repo with the tutorial's code
- Follow up for problem solving using CodersCrowd
## What environment should we use ?

Environment should be specific to each analysis. For our study group, we will be using `virtualenv` and I will show you how we can even exchange these environment with other developers
## What should we know a priori for this hacky hour ?

A little bit of python, an object oriented spirit and some cookies and coffee
## What should we be able to do after this course ?

Something similar to this :

**allele frequencies clustering (positions vs samples)**

![allele frequencies](https://bytebucket.org/radaniba/targeted-sequencing-analytics-suite/raw/c66ec81fde737c7c5159611390cd892a9307bf3f/example_output/allele_frequencies.png?token=a91e78b2141bfdc8a531dd01647de9f89b30fab8)

**coverage across positions violin plot**

![coverage](https://bytebucket.org/radaniba/targeted-sequencing-analytics-suite/raw/c66ec81fde737c7c5159611390cd892a9307bf3f/example_output/coverage_across_positions_violin.png?token=eca1c3c58643d998a2ed42f0091481e9f8bdbade)

**target capture CDF**

![capture](https://bytebucket.org/radaniba/targeted-sequencing-analytics-suite/raw/c66ec81fde737c7c5159611390cd892a9307bf3f/example_output/target_capture_cdf.png?token=326eb8883929ba10db87f4a69f5b6197c72642bb)

**Zygosity matrix**

![zygosity](https://bytebucket.org/radaniba/targeted-sequencing-analytics-suite/raw/c66ec81fde737c7c5159611390cd892a9307bf3f/example_output/zygosity_matrix.png?token=7e7564911de748ccf976476469bfd3d806a0a4d4)

**Coverage heatmap across amplicon regions**

![coverage in amplicon](https://bytebucket.org/radaniba/targeted-sequencing-analytics-suite/raw/e407a493eca27b8b8351cb8e422a7352d524f5a6/example_output/coverage_heatmap_per_amplicon.png?token=ef2eef0f330040f7fe15e094ac37db51009aee44)

**In target vs Out of targets**

![in_out](https://bytebucket.org/radaniba/targeted-sequencing-analytics-suite/raw/c66ec81fde737c7c5159611390cd892a9307bf3f/example_output/in_out_of_target.png?token=7e3b9f61b5ea40fd115502c540a3972a2f0d80c2)

**Mapped Unmapped Reads**

![mapped unmapped](https://bytebucket.org/radaniba/targeted-sequencing-analytics-suite/raw/c66ec81fde737c7c5159611390cd892a9307bf3f/example_output/mapped_reads.png?token=523e92eccb47afde3fd5b35a6e7aed374e5362d7)

**Mapping Qualities**

![mapping quality](https://bytebucket.org/radaniba/targeted-sequencing-analytics-suite/raw/c66ec81fde737c7c5159611390cd892a9307bf3f/example_output/mapping_quality.png?token=1ddbdd03613efea77b83f505e272b4ea6117d86c)

**Comment by jstaf** [comment]
This would be great! I'm really looking forward to learning how to handle DNA sequencing data (all my experience is with RNA-Seq). 

Out of curiosity, what are you using to map and process the reads? The reason I ask is that I'm anticipating WIndows users having compatibility issues with UNIX-only parts of the pipeline.

**Comment by radaniba** [comment]
@kazi11 I will be using `bowtie2` and this will be part of `wrap publically available tools` section. That said it can be `bwa` as well or any short read aligner.
I never used these tools on Windows, but I don't expect issues with it. That will be a nice test as well for reproducibility :)
We can use Docker on Windows too, that way we can all use the same OS !

**Comment by jstaf** [comment]
Huh, so I just checked the `bowtie2` documentation and it totally runs on Windows. `Tophat2` (an RNA-seq aligner that uses `bowtie2`) doesn't, so I assumed that `bowtie2` had the same problem. Apparently not (false alarm!).

**Comment by radaniba** [comment]
If you have the right compiler and the source code I guess it should work easily on windows (sometimes easier to say than to do when it comes to compiling source codes )

**Comment by katieannemills** [comment]
Hey @radaniba,

This looks awesome! Couple of comments:
- I see you've got version control listed as 1/3 of one lesson; we spent a whole hour on this just a few weeks ago. Is it possible to teach this lesson without git? You know I'm a huge fan of git, but maybe best to tackle one thing at a time, and keep the focus on the main material.
- This looks (to me) like a lot of heavy machinery - what's the plan to help as many people as possible participate? It'd be good to start by identifying all the dependencies we'll need, and figure out what we can do to smooth installation / setup (kind of sounds like you already have a plan though :)
- This is pretty python heavy; I'm a huge python fan, but I think a lot of people come particularly for the R (though I could be wrong! If anyone wants to do some python, say so here!). Would it be possible to explore these same ideas in R?
- This looks really genomics specific (which is great, there are tons of genomics people in the group!); will non-genomics people get much out of this? I'd be delighted to have a series of genomics-targeted events, but I just want to make sure we communicate that to people if that's the case, so that everyone knows what they're getting into.

**Comment by radaniba** [comment]
Hi @BillMills ,

Thanks for your comments 

> I see you've got version control listed as 1/3 of one lesson; we spent a whole hour on this just a few weeks ago. Is it possible to teach this lesson without git? You know I'm a huge fan of git, but maybe best to tackle one thing at a time, and keep the focus on the main material.

You're right, I agree, if this will save us precious time, that would be great, there is a lot of material in there and I am aware it takes a lot of time to finish that kind of program

> This looks (to me) like a lot of heavy machinery - what's the plan to help as many people as possible participate? It'd be good to start by identifying all the dependencies we'll need, and figure out what we can do to smooth installation / setup (kind of sounds like you already have a plan though :)

Yes, I planned to allocate some time to make participants install the dependencies, not a big deal though, the usual process, pip and virtualenv, the entire process will be done at 10 min or so

> This is pretty python heavy; I'm a huge python fan, but I think a lot of people come particularly for the R (though I could be wrong! If anyone wants to do some python, say so here!). Would it be possible to explore these same ideas in R?

That would be great (besides this is a subject of a new issue I will open here). That said, I always wrote pipelines in python and I am not aware of R frameworks that are equivalent to the make-like family of utilities that make it possible to 'easily' develop a pipeline (the whole purpose is making participants familiar with automation). If anyone is aware of R packages to make pipelines, please let me know.

> This looks really genomics specific (which is great, there are tons of genomics people in the group!); will non-genomics people get much out of this? I'd be delighted to have a series of genomics-targeted events, but I just want to make sure we communicate that to people if that's the case, so that everyone knows what they're getting into.

Yes it is (my daily routine). This is applied but the protocol can fit any other application. I would be more than happy to collaborate with someone from another field and do the same (even though it will be out of my comfort zone), this will demonstrate that the process is reproducible at some extent (any idea's welcome, I love challenges :) )

I agree with you, this is a bit too specific, why don't we make this a final step of some intermediate shorter session, I am thinking of : 
- Writing simple pipelines in Python
- Data analysis with R and Python together
- Data visualization in R and Python

We can break into smaller units before attacking such heavy processes 

Does this make sense ?

PS : see [my next issue](https://github.com/minisciencegirl/studyGroup/issues/23)

**Comment by jstaf** [comment]
Yeah, I don't think there are any R frameworks for pipeline programming. But hey, I really want to learn how to throw together a pipeline in Python (could be a really nice alternative to the rather finicky shell scripts I've been writing).

And @BillMills, I think there would be a lot of useful information here even for non-genomics peeps. Writing pipelines has a lot of utility for a large number of fields. Genomics is just a really good example to learn from.

**Comment by minisciencegirl** [comment]
Is it possible to throw together a Python pipeline without basic Python knowledge?

**Comment by radaniba** [comment]
Yes, with a python crash course this is totally doable

**Comment by minisciencegirl** [comment]
Cool that sounds great! Really interested in Rython tool kit as well. Let's schedule a session soon?

**Comment by katieannemills** [comment]
So how about these sessions - could be a good opportunity to do some events at BC Cancer.

**Comment by radaniba** [comment]
Oh that would be great, may be @minisciencegirl can check with VanBug team ( @minisciencegirl are you part of their core team ? ) they have easy access to facilities here so it can be a good start

**Comment by minisciencegirl** [comment]
Hi Rad,

I am on their Dev team. I can check with Kirin and Will about booking rooms. Let me know a time that works and I can get the ball rolling. Would it be possible to schedule these for later in the afternoon? 

Cheers,

Amy
